modules/support/src/bounded.rs

```diff
@@ -0,0 +1,167 @@
+// This file is part of Acala.
+
+// Copyright (C) 2022 Acala Foundation.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use super::Rate;
+
+use codec::{Decode, Encode, MaxEncodedLen};
+use frame_support::traits::Get;
+use primitives::{Balance, BlockNumber};
+use scale_info::TypeInfo;
+use sp_runtime::{
+	traits::{CheckedSub, One, Zero},
+	FixedPointNumber, RuntimeDebug,
+};
+use sp_std::{marker::PhantomData, prelude::*, result::Result};
+
+#[cfg(feature = "std")]
+use serde::{Deserialize, Serialize};
+
+/// The bounded type errors.
+#[derive(RuntimeDebug, PartialEq, Eq)]
+pub enum Error {
+	/// The value is out of bound.
+	OutOfBound,
+	/// The change diff exceeds the max absolute value.
+	ExceedMaxChangeAbs,
+}
+
+/// An abstract definition of bounded type. The type is within the range of `Range`
+/// and while update the inner value, the max absolute value of the diff is `MaxChangeAbs`.
+/// The `Default` value is minimum value of the range.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize), serde(transparent))]
+#[derive(Encode, Decode, PartialEq, Eq, PartialOrd, Ord, Copy, Clone, TypeInfo, MaxEncodedLen, RuntimeDebug)]
+#[scale_info(skip_type_params(Range, MaxChangeAbs))]
+pub struct BoundedType<T: Encode + Decode, Range, MaxChangeAbs>(
+	T,
+	#[cfg_attr(feature = "std", serde(skip_serializing))] PhantomData<(Range, MaxChangeAbs)>,
+);
+
+impl<T: Default + Encode + Decode, Range: Get<(T, T)>, MaxChangeAbs: Get<T>> Default
+	for BoundedType<T, Range, MaxChangeAbs>
+{
+	fn default() -> Self {
+		let (min, _) = Range::get();
+		Self(min, PhantomData)
+	}
+}
+
+impl<T: Encode + Decode + CheckedSub + PartialOrd + Copy, Range: Get<(T, T)>, MaxChangeAbs: Get<T>>
```

Comment:
should implement `TryFrom` and `Into` trait

Reply:
`TryFrom` and `Into` cannot be implemented for bounded types, which conflict with impl in `core`, due to rust trait limitations https://github.com/rust-lang/rust/issues/50133